src/reflect/scala/reflect/internal/tpe/TypeComparers.scala

```diff
@@ -365,9 +365,23 @@ trait TypeComparers {
 
   // @assume tp1.isHigherKinded || tp2.isHigherKinded
   def isHKSubType(tp1: Type, tp2: Type, depth: Depth): Boolean = {
-    def isSub(ntp1: Type, ntp2: Type) = (ntp1.withoutAnnotations, ntp2.withoutAnnotations) match {
+    def isSub(tp1: Type, tp2: Type) = (tp1.withoutAnnotations, tp2.withoutAnnotations) match {
       case (TypeRef(_, AnyClass, _), _)                                     => false                    // avoid some warnings when Nothing/Any are on the other side
       case (_, TypeRef(_, NothingClass, _))                                 => false
+      case (tv1 @ TypeVar(_, _), tv2 @ TypeVar(_, _)) if settings.YhkTypevarUnification =>
+        (tv2.params corresponds tv1.params)(methodHigherOrderTypeParamsSubVariance) &&
+        { tv1.addHiBound(tv2); tv2.addLoBound(tv1); true }
+      case (tp1, tv2 @ TypeVar(_, _)) if settings.YhkTypevarUnification =>
+        val ntp1 = tp1.normalize
+        (tv2.params corresponds ntp1.typeParams)(methodHigherOrderTypeParamsSubVariance) &&
+        { tv2.addLoBound(ntp1); true }
+      case (tv1 @ TypeVar(_, _), tp2) if settings.YhkTypevarUnification =>
+        val ntp2 = tp2.normalize
+        (ntp2.typeParams corresponds tv1.params)(methodHigherOrderTypeParamsSubVariance) &&
+        { tv1.addHiBound(ntp2); true }
+      case (tp1, tp2) => isSub2(tp1.normalize, tp2.normalize)  // @M! normalize reduces higher-kinded case to PolyType's
+    }
```

Comment:
The implementations of `<:<` and `=:=` usually follow the same pattern. I wonder if a similar change needed in:

```scala
  private def isSameHKTypes(tp1: Type, tp2: Type) = (
       tp1.isHigherKinded
    && tp2.isHigherKinded
    && (tp1.normalize =:= tp2.normalize)
  )
```

It would be great to setup a unit test in the same vein as https://github.com/scala/scala/blob/9acab45aeeadef2f63da69faf81465cc15599789/test/junit/scala/reflect/internal/TypesTest.scala  to directly demonstrate the new unification (rather than indirectly via compilation), and to explore whether the change preserves transitivity of subtyping and consistency with `=:=`.

Reply:
> I wonder if a similar change needed in:
>
> ```scala
> private def isSameHKTypes(tp1: Type, tp2: Type)
> ```

If `=:=` also performs unification as a side effect, then yes. Otherwise, I think it is not necessary.

Reply:
It does, via `mutateNonTypeConstructs` and then `registerTypeEquality`

Reply:
Yeah. Nevertheless, `mutateNonTypeConstructs` already seems to handle the higher-kinded `TypeVar` case correctly.